From 45ff75351b073029ec5d8261c7a85c914cb85b7f Mon Sep 17 00:00:00 2001
From: Benjamin Walsh <benjamin.walsh@windriver.com>
Date: Thu, 27 May 2010 17:38:10 -0400
Subject: [PATCH 4/5] MIPS/kexec: Add board-specific cache flushing

Add support for board-specific call for doing cache flushing on all
CPUs before jumping to the assembly code.

Signed-off-by: Benjamin Walsh <benjamin.walsh@windriver.com>
---
 arch/mips/kernel/machine_kexec.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/arch/mips/kernel/machine_kexec.c b/arch/mips/kernel/machine_kexec.c
index c1fddbb..cbdb9be 100644
--- a/arch/mips/kernel/machine_kexec.c
+++ b/arch/mips/kernel/machine_kexec.c
@@ -22,6 +22,7 @@ extern unsigned long kexec_indirection_page;
 int (*_machine_kexec_prepare)(struct kimage *);
 void (*_machine_kexec_shutdown)(void);
 void (*_machine_crash_shutdown)(struct pt_regs *regs);
+void (*_machine_cache_flush)(void) = NULL;
 void (*_machine_smp_handle_restart)(unsigned long reloc) = NULL;
 
 int
@@ -122,7 +123,12 @@ machine_kexec(struct kimage *image)
 
 	printk("Will call new kernel at %08lx\n", image->start);
 	printk("Bye ...\n");
-	__flush_cache_all();
+
+	if(_machine_cache_flush) {
+		_machine_cache_flush();
+	} else {
+		__flush_cache_all();
+	}
 
 	if(_machine_smp_handle_restart) {
 		_machine_smp_handle_restart(reboot_code_buffer);
-- 
1.6.5.2

